src/packet_builder.rs

```diff
@@ -464,6 +464,34 @@ impl PacketBuilderStep<VlanHeader> {
 }
 
 impl PacketBuilderStep<IpHeader> {
+    pub fn icmp4(mut self, icmp_type: IcmpV4Type, icmp_code: u8) -> PacketBuilderStep<IcmpV4Header> {
+        self.state.transport_header = Some(TransportHeader::Icmp4(IcmpV4Header{
+            icmp_type,
+            icmp_code,
+            icmp_chksum: 0, // calculated later
+            echo_header: None,
+        }));
+        //return for next step
+        PacketBuilderStep {
+            state: self.state,
+            _marker: marker::PhantomData::<IcmpV4Header>{}
+        }
+    }
+
+    pub fn icmp6(mut self, icmp_type: IcmpV6Type, icmp_code: u8) -> PacketBuilderStep<IcmpV6Header> {
+        self.state.transport_header = Some(TransportHeader::Icmp6(IcmpV6Header{
+            icmp_type,
+            icmp_code,
+            icmp_chksum: 0, // calculated later
+            echo_header: None,
+        }));
+        //return for next step
+        PacketBuilderStep {
+            state: self.state,
+            _marker: marker::PhantomData::<IcmpV6Header>{}
+        }
+    }
+
```

Comment:
I would remove the echo from the next step and instead add four methods for the echo case directly:

* `icmp4_echo_reply(seq, id)`
* `icmp6_echo_request(seq, id)`
* `icmp6_echo_reply(seq, id)`
* `icmp6_echo_reply(seq, id)`

that way it is no needed to specify the type & code as they are implicitly known from the type (also plays nicely with the `content` comment I wrote for the IcmpV4Header & IcmpV6Header). You also don't need a sanity check.

But I would keep the `icmp4` & `icmp6` methods so it is possible to write arbitrary icmp packets (if the user is so inclined).

Then creating an ICMP echo replay packet would then look like this:

```
        let builder = PacketBuilder::
             ipv4(   [192,168,1,1],  //source ip
                 [192,168,1,2], //desitionation ip
                 20)            //time to life
              .icmp4_echo_request(1, 2)
```